pubtools/_pulp/tasks/delete.py

```diff
@@ -0,0 +1,709 @@
+
+        missing = set(repos) - set(repo_map.keys())
+        if missing:
+            missing = ", ".join(sorted(list(missing)))
+            LOG.warning("No units to remove from %s", missing)
+
+        return repo_map, unit_map
+
+    @step("Remove RPMs")
+    def remove_rpms(self, repo_map):
+        return self.delete_content(RpmUnit, repo_map, self._rpm_remove_crit)
+
+    @step("Remove ISOs")
+    def remove_files(self, repo_map):
+        return self.delete_content(FileUnit, repo_map, self._file_remove_crit)
+
+    @step("Remove modules")
+    def remove_modules(self, repo_map):
+        return self.delete_content(ModulemdUnit, repo_map, self._module_remove_crit)
+
+    def delete_content(self, unit_type, repo_map, criteria_fn):
+        out = []
+        # get repos
+        repos = self.search_repo(repo_map.keys())
+
+        # request removal
+        for repo in sorted(repos):
+            units = repo_map.get(repo.id)
+            criteria = self.unit_criteria(unit_type, criteria_fn(units))
+            f = repo.remove_content(criteria=criteria)
+            f = f_map(f, partial(ClearedRepo, repo=repo, content=units))
+            f = f_map(f, self.log_remove)
+            out.append(f)
+
+        return out
+
+    def _rpm_remove_crit(self, units):
+        part_crit = []
+        for unit in units:
+            part_crit.append(self._rpm_criteria(unit.filename, [unit.signing_key]))
+
+        return part_crit
+
+    def _file_remove_crit(self, units):
+        part_crit = []
+        for unit in units:
+            part_crit.append(self._file_criteria(unit.path))
+
+        return part_crit
+
+    def _module_remove_crit(self, units):
+        part_crit = []
+        for unit in units:
+            part_crit.append(self._module_criteria(unit.nsvca))
+
+        return part_crit
+
+    def search_repo(self, repo_ids):
+        return self.pulp_client.search_repository(Criteria.with_id(repo_ids)).result()
+
+    @step("Remove packages in advisory")
+    def remove_adv_items(self, repo_map):
+        out = []
+        # get repos from pulp
+        repos = self.search_repo(repo_map.keys())
+
+        for repo in repos:
+            items = repo_map.get(repo.id)
+            f = self.remove_adv_content(repo, items)
+            f = f_map(f, partial(ClearedRepo, repo=repo, content=items))
+            f = f_map(f, self.log_remove)
+            out.append(f)
+
+        return out
+
+    def remove_adv_content(self, repo, items):
+        crit = []
+        for item in items:
+            crit.append(
+                Criteria.and_(
+                    Criteria.with_field("filename", item.name),
+                    Criteria.with_field("signing_key", item.signing_key.lower()),
+                )
+            )
+
+        criteria = Criteria.and_(Criteria.with_unit_type(RpmUnit), Criteria.or_(*crit))
+
+        return repo.remove_content(criteria=criteria)
+
+    @step("Sort out files")
```

Comment:
I don't think this should be a step either, it should complete more or less instantaneously.

Reply:
Ack.

Reply:
Removed it